### ISSUE TITLE
language code for translations

### ISSUE THREAD

**Issue by elarlang** [OWASP/ASVS#1175]
We should pick, document and implement, what is our policy for language codes for translations - like in file names, folder names etc.

Option 1 - use correct 2 letter language codes based on ISO-639-1:
<https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes>

Option 2 - we can be more precise with language (locale) code, like instead of "en" we should use "en-gb" or "en-us". Then based on that it's possible to make more precise language corrections in documents (and use correct speller). The same may apply for Portuguese (pt-pt, pt-br), Spanish (es-es, es-*), Chinese (zh-cn, zh-*), Arabic (ar-*) etc.

In other words - use "Language-Region" from RFC5646 (<https://datatracker.ietf.org/doc/html/rfc5646>).

There are some other combinations available as well, so we need to pick one and use only this one.

<https://www.w3.org/International/articles/language-tags/>

**Comment by danielcuthbert** [comment]
I would prefer to be more precise, I mean US English versus UK English is enough to cause conflict. I'd say we adopt en-gb as you said, which should be more useful to all

**Comment by tghosth** [comment]
Is the benefit from trying to enforce this worth the change. Right now, things seem to work ok with 2 letter codes and we don't have so many translations that it is a problem.

**Comment by elarlang** [comment]
It is kind of time to decide this for ASVS v5.0 translations.

**Comment by randomstuff** [comment]
> In other words - use "Language-Region" from RFC5646 (https://datatracker.ietf.org/doc/html/rfc5646).

Yes, I would use that. (Maybe someone will make a en-GB translation!)

**Comment by tghosth** [comment]
So I am in favor of keeping two letter codes (like [this](https://en.wikipedia.org/wiki/List_of_ISO_639_language_codes)) for a few reasons:

1. Simpler
2. I don't think we will get that many translations. Realistically no one is translating into en-gb 🤣😂🤣.
3. Using something else will break the output scripts.

**Comment by elarlang** [comment]
I don't think we will have en-gb and en-us, but we may have different versions of Chinese, Spanish, or Portuguese.

> The same may apply for Portuguese (pt-pt, pt-br), Spanish (es-es, es-*), Chinese (zh-cn, zh-*), Arabic (ar-*) etc.

I agree, that 2-letter version is nicer, but it may cause limits.

**Comment by jmanico** [comment]
ISO 639-1 (two character code) covers about 180 languages. What does this miss?

**Comment by elarlang** [comment]
```
  'es': 'Spanish',
```

vs
```
  'es-AR': 'Spanish (Argentina)',
  'es-BO': 'Spanish (Bolivia)',
  'es-CL': 'Spanish (Chile)',
  'es-CO': 'Spanish (Colombia)',
  'es-CR': 'Spanish (Costa Rica)',
  'es-DO': 'Spanish (Dominican Republic)',
  'es-EC': 'Spanish (Ecuador)',
  'es-ES': 'Spanish (Spain)',
  'es-GT': 'Spanish (Guatemala)',
  'es-HN': 'Spanish (Honduras)',
  'es-MX': 'Spanish (Mexico)',
  'es-NI': 'Spanish (Nicaragua)',
  'es-PA': 'Spanish (Panama)',
  'es-PE': 'Spanish (Peru)',
  'es-PR': 'Spanish (Puerto Rico)',
  'es-PY': 'Spanish (Paraguay)',
  'es-SV': 'Spanish (El Salvador)',
  'es-UY': 'Spanish (Uruguay)',
  'es-VE': 'Spanish (Venezuela)',
```

**Comment by randomstuff** [comment]
> ISO 639-1 (two character code) covers about 180 languages. What does this miss?

Klingon 😄

Actually we can probably start with RFC5646 primary language subtags (en, fr, es, zh, and tlh). If someone ever feels the need to have different variants it will be possible to split eg. "fr" into "fr-FR", "fr-CA", "fr-BE", "fr-CF" and more when needed.

**Comment by jmanico** [comment]
> > ISO 639-1 (two character code) covers about 180 languages. What does this miss?
> 
> Klingon 😄
> 
> Actually we can probably start with RFC5646 primary language subtags (en, fr, es, zh, and tlh). If someone ever feels the need to have different variants it will be possible to split eg. "fr" into "fr-FR", "fr-CA", "fr-BE", "fr-CF" and more when needed.

Klingon is TL as a 2-letter code. ISO 639-3 and ISO 639-2 consider Klingon to be TLH.

Live long and prosper, @randomstuff
